src/vm/checker/typecheck/natives/assets.rs

```diff
@@ -0,0 +1,166 @@
+    let expected_asset_type = checker.contract_context.get_asset_type(asset_name)
+        .ok_or(CheckErrors::NoSuchAsset(asset_name.clone()))?;
+
+    if !expected_asset_type.admits_type(&supplied_asset_type) {
+        return Err(CheckErrors::TypeError(expected_asset_type.clone(), supplied_asset_type).into())
+    }
+
+    if !expected_owner_type.admits_type(&supplied_owner_type) {
+        return Err(CheckErrors::TypeError(expected_owner_type, supplied_owner_type).into())
+    }
+
+    Ok(AtomTypeIdentifier::ResponseType(
+        Box::new((AtomTypeIdentifier::BoolType.into(),
+                  AtomTypeIdentifier::IntType.into()))).into())
+}
+
+pub fn check_special_mint_token(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(3, args)?;
+
+    let asset_name = args[0].match_atom()
+        .ok_or(CheckErrors::BadAssetName)?;
+    checker.type_map.set_type(&args[0], no_type())?;
+
+    let supplied_amount = checker.type_check(&args[1], context)?;
+    let supplied_owner_type = checker.type_check(&args[2], context)?;
+
+    let expected_amount: TypeSignature = AtomTypeIdentifier::IntType.into();
+    let expected_owner_type: TypeSignature = AtomTypeIdentifier::PrincipalType.into();
+
+    if !checker.contract_context.token_exists(asset_name) {
+        return Err(CheckErrors::NoSuchAsset(asset_name.clone()).into());
+    }
+
+    if !expected_amount.admits_type(&supplied_amount) {
+        return Err(CheckErrors::TypeError(expected_amount, supplied_amount).into())
+    }
+
+    if !expected_owner_type.admits_type(&supplied_owner_type) {
+        Err(CheckErrors::TypeError(expected_owner_type, supplied_owner_type).into())
+    } else {
+        Ok(AtomTypeIdentifier::ResponseType(
+            Box::new((AtomTypeIdentifier::BoolType.into(),
+                      AtomTypeIdentifier::IntType.into()))).into())
+    }
+}
+
+pub fn check_special_transfer_asset(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(4, args)?;
+
+    let asset_name = args[0].match_atom()
+        .ok_or(CheckErrors::BadAssetName)?;
+    checker.type_map.set_type(&args[0], no_type())?;
+
+    let supplied_asset_type = checker.type_check(&args[1], context)?;
+    let supplied_owner_type = checker.type_check(&args[2], context)?;
+    let supplied_recipient_type = checker.type_check(&args[3], context)?;
+
+    let expected_owner_type: TypeSignature = AtomTypeIdentifier::PrincipalType.into();
+    let expected_asset_type = checker.contract_context.get_asset_type(asset_name)
+        .ok_or(CheckErrors::NoSuchAsset(asset_name.clone()))?;
+
+    if !expected_asset_type.admits_type(&supplied_asset_type) {
+        return Err(CheckErrors::TypeError(expected_asset_type.clone(), supplied_asset_type).into())
+    }
+
+    if !expected_owner_type.admits_type(&supplied_owner_type) {
```

Comment:
Should we have different error types for whether or not the owner or recipient types are invalid?

Reply:
Probably not -- it would probably be better if the error specified the offending argument's SymbolicRepresentation (that would specify exactly which argument caused the type error).

Reply:
Sounds good to me!

Reply:
Addressed in 457c403c -- generalized this to a function in the checker. Will open a separate PR to use that function for other instances.